### PR TITLE
support record creation with externalUserId

### LINKED ISSUE #98: Support record creation with externalUserId
As we discussed, not all data uploads are meant to know the user-id(subject-id) and it is hard to manually enter on a smart phone. So it would be nice to add support to enter external-id and let the backend fetch the user-id internally.
This can also include a change on the UI where they can enter either user-id or external-id when creating a record
@blootsvoets @khoathanglong

### PR DESCRIPTION
**Pull request by nivemaham**
support record creation with externalUserId
add variables to configure intervals to sync with MP
Fixes #98

### CODE REVIEW

File: radar-upload-backend/src/main/java/org/radarbase/upload/Config.kt [hunk 1]
```diff
@@ -42,4 +42,6 @@ data class Config(
         var sourceTypes: List<SourceTypeDTO>? = null,
         var additionalPersistenceConfig: Map<String, String>? = null,
         var contentStreamBufferSize: Long = 1048576L,
-        var enableCors: Boolean? = false)
+        var enableCors: Boolean? = false,
+        var syncProjects: Long? = 30,
```
Comment (blootsvoets) [inline radar-upload-backend/src/main/java/org/radarbase/upload/Config.kt]: rename to `syncProjectsIntervalMin`? Similarly for `syncParticipants`. I would suggest to make this a non-null field, so either this default is used or any user-provided value. This eliminates non-null coercion (`!!`) when evaluating the variables.

File: radar-upload-backend/src/test/java/org/radarbase/upload/mock/MockResourceEnhancerFactory.kt [hunk 2]
```diff
@@ -59,25 +60,32 @@ class MockResourceEnhancerFactory(
...
 
-        override fun userProjects(auth: Auth) = projects.keys.map { Project(id = it) }
+}
 
-        override fun projectUsers(projectId: String): List<User> {
-            ensureProject(projectId)
-            return projects.getValue(projectId)
-                    .map { User(projectId = projectId, id = it, status = "ACTIVATED") }
-        }
+class MockProjectService(private val projects: Map<String, List<String>>) : UploadProjectService {
```
Comment (blootsvoets) [inline radar-upload-backend/src/test/java/org/radarbase/upload/mock/MockResourceEnhancerFactory.kt]: If this is now an upper level class, you could move it to a separate file.

File: radar-upload-backend/src/test/java/org/radarbase/upload/resource/RecordResourceTest.kt [hunk 3]
```diff
@@ -27,6 +27,7 @@ import org.radarbase.upload.Config
 import org.radarbase.upload.api.RecordDTO
 import org.radarbase.upload.api.RecordDataDTO
 import org.radarbase.upload.api.SourceTypeDTO
+import org.radarbase.upload.logger
```
Comment (blootsvoets) [inline radar-upload-backend/src/test/java/org/radarbase/upload/resource/RecordResourceTest.kt]: not used?

File: radar-upload-backend/src/main/java/org/radarbase/upload/service/managementportal/MPProjectService.kt [hunk 4]
```diff
@@ -56,11 +57,15 @@ class MPProjectService(@Context private val mpClient: MPClient): UploadProjectSe
...
+            CachedSet(Duration.ofMinutes(config.syncParticipants!!), Duration.ofMinutes(1)) {
                 mpClient.readParticipants(projectId)
             }
         }
 
         return projectParticipants.get().toList()
     }
+
+    override fun userByExternalId(projectId: String, externalUserId: String): User? =
+            projectUsers(projectId).firstOrNull { it.externalId == externalUserId }
```
Comment (blootsvoets) [inline radar-upload-backend/src/main/java/org/radarbase/upload/service/managementportal/MPProjectService.kt]: I would suggest to do a `find` here instead, and throw an `IllegalStateException` if there is more than one user with the same externalUserId. Alternatively, make this method return a List<User> and throw an exception in `RecordMapperImpl.resolveUserId`

Just to check: how does ManagementPortal handle duplicate external IDs? Is it forbidden?

Comment (nivemaham) [reply radar-upload-backend/src/main/java/org/radarbase/upload/service/managementportal/MPProjectService.kt]: ManagmentPortal does not allow duplicate externalIds within a project.

File: radar-upload-backend/src/main/java/org/radarbase/upload/api/RecordMapperImpl.kt [hunk 5]
```diff
@@ -53,6 +55,18 @@ class RecordMapperImpl(
         return Pair(recordDoa, toMetadata(record.metadata))
     }
 
+    private fun RecordDataDTO.resolveUserId(): String {
+        when (this.userId != null) {
```
Comment (blootsvoets) [inline radar-upload-backend/src/main/java/org/radarbase/upload/api/RecordMapperImpl.kt]: why not use
```kotlin
return if (userId != null) userId!! else {
     ...
}
```
or reduce the when statement to
```kotlin
return when {
   userId != null -> userId
   externalUserId != null -> { ... }
   else -> throw BadRequestException("Missing both user ID and external-user ID")
}
```

**Pull request merged**